File: app/src/main/java/org/dhis2/usescases/datasets/datasetDetail/DataSetDetailActivity.java
package org.dhis2.usescases.datasets.datasetDetail;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.transition.ChangeBounds;
import android.transition.Transition;
import android.transition.TransitionManager;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.core.view.ViewCompat;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import org.dhis2.App;
import org.dhis2.R;
import org.dhis2.bindings.ExtensionsKt;
import org.dhis2.bindings.ViewExtensionsKt;
import org.dhis2.commons.Constants;
import org.dhis2.commons.filters.FilterItem;
import org.dhis2.commons.filters.FilterManager;
import org.dhis2.commons.filters.FiltersAdapter;
import org.dhis2.commons.orgunitselector.OUTreeFragment;
import org.dhis2.commons.sync.ConflictType;
import org.dhis2.databinding.ActivityDatasetDetailBinding;
import org.dhis2.ui.ThemeManager;
import org.dhis2.usescases.datasets.datasetDetail.datasetList.DataSetListFragment;
import org.dhis2.usescases.general.ActivityGlobalAbstract;
import org.dhis2.utils.DateUtils;
import org.dhis2.utils.category.CategoryDialog;
import org.dhis2.commons.sync.SyncContext;
import org.dhis2.utils.granularsync.SyncStatusDialog;
import org.dhis2.utils.granularsync.SyncStatusDialogNavigatorKt;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit;

import java.util.List;

import javax.inject.Inject;

import dhis2.org.analytics.charts.ui.GroupAnalyticsFragment;
import kotlin.Unit;


public class DataSetDetailActivity extends ActivityGlobalAbstract implements DataSetDetailView {

    private ActivityDatasetDetailBinding binding;
    private String dataSetUid;
    public DataSetDetailComponent dataSetDetailComponent;

    private Fragment fragment;

    @Inject
    DataSetDetailPresenter presenter;

    @Inject
    FilterManager filterManager;

    @Inject
    FiltersAdapter filtersAdapter;

    @Inject
    ThemeManager themeManager;

    @Inject
    DataSetDetailViewModelFactory viewModelFactory;

    private boolean backDropActive;

    private DataSetDetailViewModel viewModel;

    public static Intent intent(
            @NonNull Context context,
            @NonNull String dataSetUid) {
        Intent intent = new Intent(context, DataSetDetailActivity.class);
        intent.putExtra("DATASET_UID", dataSetUid);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        dataSetUid = getIntent().getStringExtra("DATASET_UID");
        dataSetDetailComponent = ((App) getApplicationContext()).userComponent().plus(new DataSetDetailModule(this, dataSetUid));
        dataSetDetailComponent.inject(this);
        themeManager.setDataSetTheme(dataSetUid);
        super.onCreate(savedInstanceState);

        viewModel = new ViewModelProvider(this, viewModelFactory).get(DataSetDetailViewModel.class);

        binding = DataBindingUtil.setContentView(this, R.layout.activity_dataset_detail);
        binding.setName(getIntent().getStringExtra(Constants.DATA_SET_NAME));
        binding.setPresenter(presenter);

        ViewExtensionsKt.clipWithRoundedCorners(binding.eventsLayout, ExtensionsKt.getDp(16));
        binding.filterLayout.setAdapter(filtersAdapter);
        configureBottomNavigation();

        if (SyncStatusDialogNavigatorKt.shouldLaunchSyncDialog(getIntent())) {
            showGranularSync();
        }
    }

    private void configureBottomNavigation() {
        boolean accessWriteData = Boolean.parseBoolean(getIntent().getStringExtra(Constants.ACCESS_DATA));
        viewModel.getPageConfiguration().observe(this, binding.navigationBar::pageConfiguration);

        binding.navigationBar.setOnItemSelectedListener(item -> {
            Fragment newFragment = null;
            switch (item.getItemId()) {
                case R.id.navigation_list_view:
                    newFragment = DataSetListFragment.newInstance(dataSetUid, accessWriteData);
                    break;
                case R.id.navigation_analytics:
                    presenter.trackDataSetAnalytics();
                    newFragment = GroupAnalyticsFragment.Companion.forDataSet(dataSetUid);
                    break;
            }
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            if (fragment == null ||(newFragment != null  && !fragment.getClass().toString().equals(newFragment.getClass().toString()))) {
                fragment = newFragment;
                transaction.replace(R.id.fragmentContainer, fragment).commit();
            }
            return true;
        });
        binding.navigationBar.selectItemAt(0);
        binding.fragmentContainer.setPadding(0, 0, 0, binding.navigationBar.isHidden() ? 0 : ExtensionsKt.getDp(56));

    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.init();
        binding.setTotalFilters(FilterManager.getInstance().getTotalFilters());
    }

    @Override
    protected void onPause() {
        presenter.setOpeningFilterToNone();
        presenter.onDettach();
        super.onPause();
    }

    @Override
    public void showHideFilter() {
        Transition transition = new ChangeBounds();
        transition.setDuration(200);
        TransitionManager.beginDelayedTransition(binding.backdropLayout, transition);
        backDropActive = !backDropActive;
        ConstraintSet initSet = new ConstraintSet();
        initSet.clone(binding.backdropLayout);
        if (backDropActive) {
            initSet.connect(R.id.eventsLayout, ConstraintSet.TOP, R.id.filterLayout, ConstraintSet.BOTTOM, 50);
        } else {
            initSet.connect(R.id.eventsLayout, ConstraintSet.TOP, R.id.backdropGuideTop, ConstraintSet.BOTTOM, 0);
        }
        initSet.applyTo(binding.backdropLayout);

        binding.filterOpen.setVisibility(backDropActive ? View.VISIBLE : View.GONE);
        ViewCompat.setElevation(binding.eventsLayout, backDropActive ? 20 : 0);
    }

    @Override
    public void clearFilters() {
        filtersAdapter.notifyDataSetChanged();
    }

    @Override
    public void updateFilters(int totalFilters) {
        binding.setTotalFilters(totalFilters);
        binding.executePendingBindings();
    }

    @Override
    public void openOrgUnitTreeSelector() {
        new OUTreeFragment.Builder()
                .showAsDialog()
                .withPreselectedOrgUnits(FilterManager.getInstance().getOrgUnitUidsFilters())
                .onSelection(selectedOrgUnits -> {
                    presenter.setOrgUnitFilters((List<OrganisationUnit>) selectedOrgUnits);
                    return Unit.INSTANCE;
                })
                .build()
                .show(getSupportFragmentManager(), "OUTreeFragment");
    }

    @Override
    public void showPeriodRequest(FilterManager.PeriodRequest periodRequest) {
        if (periodRequest == FilterManager.PeriodRequest.FROM_TO) {
            DateUtils.getInstance().fromCalendarSelector(this, datePeriods -> filterManager.addPeriod(datePeriods));
        } else {
            DateUtils.getInstance().showPeriodDialog(
                    this,
                    datePeriods -> filterManager.addPeriod(datePeriods),
                    true
            );
        }
    }

    @Override
    public void showCatOptComboDialog(String catComboUid) {
        new CategoryDialog(
                CategoryDialog.Type.CATEGORY_OPTION_COMBO,
                catComboUid,
                false,
                null,
                selectedCatOptionCombo -> {
                    presenter.filterCatOptCombo(selectedCatOptionCombo);
                    return null;
                }
        ).show(
                getSupportFragmentManager(),
                CategoryDialog.Companion.getTAG()
        );
    }

    @Override
    public void setFilters(List<FilterItem> filterItems) {
        filtersAdapter.submitList(filterItems);
    }

    @Override
    public void hideFilters() {
        binding.filter.setVisibility(View.GONE);
    }

    @Override
    protected void onDestroy() {
        presenter.clearFilterIfDatasetConfig();
        super.onDestroy();
    }

    public void setProgress(boolean active) {
        binding.programProgress.setVisibility(active ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showGranularSync() {
        presenter.trackDataSetGranularSync();
        new SyncStatusDialog.Builder()
                .withContext(this, null)
                .withSyncContext(new SyncContext.DataSet(dataSetUid))
                .onDismissListener(hasChanged -> presenter.refreshList())
                .show("DATASET_SYNC");
    }
}

File: app/src/main/java/org/dhis2/usescases/datasets/datasetDetail/DataSetDetailActivity.kt
/*
 * Copyright (c) 2004-2019, University of Oslo
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of the HISP project nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.dhis2.usescases.datasets.datasetDetail

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.transition.ChangeBounds
import android.transition.TransitionManager
import android.view.View
import androidx.constraintlayout.widget.ConstraintSet
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DividerItemDecoration
import javax.inject.Inject
import org.dhis2.App
import org.dhis2.R
import org.dhis2.data.tuples.Pair
import org.dhis2.databinding.ActivityDatasetDetailBinding
import org.dhis2.usescases.datasets.dataSetTable.DataSetTableActivity
import org.dhis2.usescases.datasets.datasetInitial.DataSetInitialActivity
import org.dhis2.usescases.general.ActivityGlobalAbstract
import org.dhis2.usescases.orgunitselector.OUTreeActivity
import org.dhis2.utils.Constants
import org.dhis2.utils.DateUtils
import org.dhis2.utils.filters.FilterManager
import org.dhis2.utils.filters.FiltersAdapter
import org.dhis2.utils.granularsync.GranularSyncContracts
import org.dhis2.utils.granularsync.SyncStatusDialog
import org.hisp.dhis.android.core.category.CategoryCombo
import org.hisp.dhis.android.core.category.CategoryOptionCombo

class DataSetDetailActivity : ActivityGlobalAbstract(), DataSetDetailView {

    @Inject
    lateinit var presenter: DataSetDetailPresenter
    @Inject
    lateinit var filterManager: FilterManager

    private lateinit var binding: ActivityDatasetDetailBinding
    private lateinit var adapter: DataSetDetailAdapter
    private lateinit var filtersAdapter: FiltersAdapter

    private lateinit var dataSetUid: String
    private var accessDataWrite: Boolean = false
    private var backDropActive: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        dataSetUid = intent.getStringExtra(Constants.DATASET_UID)
        (applicationContext as App).userComponent()!!.plus(
            DataSetDetailModule(this, dataSetUid)
        ).inject(this)
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_dataset_detail)
        accessDataWrite = intent.getStringExtra(Constants.ACCESS_DATA)?.toBoolean() ?: false
        adapter = DataSetDetailAdapter(presenter)
        filtersAdapter = FiltersAdapter()

        binding.apply {
            name = intent.getStringExtra(Constants.DATA_SET_NAME)
            presenter = this@DataSetDetailActivity.presenter
            filterLayout.adapter = filtersAdapter
        }
    }

    override fun onResume() {
        super.onResume()
        presenter.init()
        binding.apply {
            addDatasetButton.isEnabled = true
            totalFilters = filterManager.totalFilters
        }
        filtersAdapter.notifyDataSetChanged()
    }

    override fun onPause() {
        presenter.onDettach()
        super.onPause()
    }

    override fun setData(dataSets: List<DataSetDetailModel>) {
        binding.programProgress.visibility = View.GONE
        if (binding.recycler.adapter == null) {
            binding.apply {
                recycler.adapter = adapter
                recycler.addItemDecoration(
                    DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
                )
            }
        }
        if (dataSets.isEmpty()) {
            binding.emptyTeis.visibility = View.VISIBLE
            binding.recycler.visibility = View.GONE
        } else {
            binding.emptyTeis.visibility = View.GONE
            binding.recycler.visibility = View.VISIBLE
            adapter.setDataSets(dataSets)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == FilterManager.OU_TREE && resultCode == Activity.RESULT_OK) {
            filtersAdapter.notifyDataSetChanged()
            updateFilters(filterManager.totalFilters)
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun showHideFilter() {
        val transition = ChangeBounds()
        transition.duration = 200
        TransitionManager.beginDelayedTransition(binding.backdropLayout, transition)
        backDropActive = !backDropActive
        val initSet = ConstraintSet()
        initSet.clone(binding.backdropLayout)

        if (backDropActive) {
            initSet.connect(
                R.id.eventsLayout,
                ConstraintSet.TOP,
                R.id.filterLayout,
                ConstraintSet.BOTTOM,
                50
            )
        } else {
            initSet.connect(
                R.id.eventsLayout,
                ConstraintSet.TOP,
                R.id.backdropGuideTop,
                ConstraintSet.BOTTOM,
                0
            )
        }
        initSet.applyTo(binding.backdropLayout)

        binding.filterOpen.visibility = if (backDropActive) View.VISIBLE else View.GONE
    }

    override fun clearFilters() {
        filtersAdapter.notifyDataSetChanged()
    }

    override fun updateFilters(totalFilters: Int) {
        binding.totalFilters = totalFilters
        binding.executePendingBindings()
    }

    override fun openOrgUnitTreeSelector() {
        val ouTreeIntent = Intent(this, OUTreeActivity::class.java)
        startActivityForResult(ouTreeIntent, FilterManager.OU_TREE)
    }

    override fun showPeriodRequest(periodRequest: FilterManager.PeriodRequest) {
        if (periodRequest == FilterManager.PeriodRequest.FROM_TO) {
            DateUtils.getInstance().showFromToSelector(this) {
                filterManager.addPeriod(it)
            }
        } else {
            DateUtils.getInstance().showPeriodDialog(
                this, { datePeriods -> filterManager.addPeriod(datePeriods) },
                true
            )
        }
    }

    override fun setCatOptionComboFilter(
        categoryOptionCombos: Pair<CategoryCombo, List<CategoryOptionCombo>>
    ) {
        filtersAdapter.addCatOptCombFilter(categoryOptionCombos)
    }

    @SuppressLint("RestrictedApi")
    override fun setWritePermission(canWrite: Boolean) {
        binding.addDatasetButton.visibility = if (canWrite) View.VISIBLE else View.GONE
    }

    override fun startNewDataSet() {
        binding.addDatasetButton.isEnabled = false
        val bundle = Bundle()
        bundle.putString(Constants.DATA_SET_UID, dataSetUid)
        startActivity(DataSetInitialActivity::class.java, bundle, false, false, null)
    }

    override fun openDataSet(dataSet: DataSetDetailModel) {
        val bundle = Bundle()
        bundle.putString(Constants.ORG_UNIT, dataSet.orgUnitUid())
        bundle.putString(Constants.ORG_UNIT_NAME, dataSet.nameOrgUnit())
        bundle.putString(Constants.PERIOD_ID, dataSet.periodId())
        bundle.putString(Constants.PERIOD_TYPE, dataSet.periodType())
        bundle.putString(Constants.PERIOD_TYPE_DATE, dataSet.namePeriod())
        bundle.putString(Constants.CAT_COMB, dataSet.catOptionComboUid())
        bundle.putString(Constants.DATA_SET_UID, dataSetUid)
        bundle.putBoolean(Constants.ACCESS_DATA, accessDataWrite)
        startActivity(DataSetTableActivity::class.java, bundle, false, false, null)
    }

    override fun showSyncDialog(dataSet: DataSetDetailModel) {
        val dialog = SyncStatusDialog.Builder()
            .setConflictType(SyncStatusDialog.ConflictType.DATA_VALUES)
            .setUid(dataSetUid)
            .setOrgUnit(dataSet.orgUnitUid())
            .setAttributeOptionCombo(dataSet.catOptionComboUid())
            .setPeriodId(dataSet.periodId())
            .onDismissListener(object : GranularSyncContracts.OnDismissListener {
                override fun onDismiss(hasChanged: Boolean) {
                    if (hasChanged) {
                        presenter.updateFilters()
                    }
                }
            })
            .build()

        dialog.show(supportFragmentManager, dialog.dialogTag)
    }
}
